### PR TITLE
[macOS/iOS] Use hardware sampling rates for audio I/O.

### LINKED ISSUE #58180: macOS audio recording error: input_callback: AudioUnitRender failed, code: -10863
### Godot version

v3.4.2

### System information

macos 11.6 GLES2

### Issue description

input_callback: AudioUnitRender failed, code: -10863
drivers/coreaudio/audio_driver_coreaudio.cpp:237 @ input_callback()

Similar problems and solutions：https://github.com/baresip/baresip/issues/601 
I hope a great god can solve this problem.

### Steps to reproduce

/Applications/Utilities/Audio\ MIDI\ Setup.app  SampleRate 48000
godot MixRate 44100

It can be reproduced by using plug-in.

### Minimal reproduction project

https://github.com/c-as/godot-voip

### PR DESCRIPTION
**Pull request by bruvzg**
Instead of "mix_rate" from the project settings, use preferred OS values, since API can work only with the limited set of values depending on device.

Fixes https://github.com/godotengine/godot/issues/58180

TODO:
- I'm not sure if changes in the `AudioStreamPlaybackMicrophone` are sufficient to correctly resample it.
- We should track device changes and adjust sampling rates accordingly, but I'm not sure if Godot audio subsystem is capable of handling sampling rate changes.

_Edit: I corrected a typo in "mix_rate". - @adamscott_

### CODE REVIEW

**Comment by kus04e4ek** [comment]
I'm still new to the Godot's source code, so take everything I say with a grain of salt.

> I'm not sure if changes in the `AudioStreamPlaybackMicrophone` are sufficient to correctly resample it.

They look like they are. `AudioStreamPlaybackMicrophone` is a child of `AudioStreamPlaybackResampled`, `AudioStreamPlaybackResampled::mix` calls `_mix_internal` and it uses `get_stream_sampling_rate` and `AudioDriver::get_mix_rate` to correctly resample audio.

> I'm not sure if Godot audio subsystem is capable of handling sampling rate changes.

Since audio generation is happening only in a callback by the current output device and uses `AudioDriver::get_mix_rate` to resample all of the audio, as long as `AudioDriver::get_mix_rate` returns the correct mix rate for the current output device and `AudioDriver::get_sample_mix_rate` returns the correct mix rate for the input device that last wrote to the buffer everything should be fine (`AudioDriverCoreAudio::output_callback`, then `AudioDriver::audio_server_process`, then `AudioServer::_driver_process`, then `AudioServer::_mix_step`, then `playback->stream_playback->mix` or `AudioStreamPlaybackResampled::mix`)

File: servers/audio_server.h [hunk 1]
```diff
@@ -97,6 +97,7 @@ class AudioDriver {
 	virtual Error init() = 0;
 	virtual void start() = 0;
 	virtual int get_mix_rate() const = 0;
+	virtual int get_capture_mix_rate() const { return get_mix_rate(); };
```
Comment (kus04e4ek) [inline servers/audio_server.h:100]: Probably `get_input_mix_rate` is a better name. See #69120

**Review by adamscott — APPROVED** [review]
Tested out the [mic record](https://github.com/godotengine/godot-demo-projects/tree/master/audio/mic_record) demo and it works flawlessly.

The code looks good too. So I'm giving it my thumbs up.

Great job, @bruvzg !

File: drivers/coreaudio/audio_driver_coreaudio.mm [hunk 2]
```diff
@@ -116,7 +116,24 @@ Error AudioDriverCoreAudio::init() {
 			break;
 	}
 
-	mix_rate = _get_configured_mix_rate();
+#if MACOS_ENABLED
```
[the remark this answers is not in the document]
Comment (adamscott) [reply drivers/coreaudio/audio_driver_coreaudio.mm:124]: Fortunately, my tests still work after fixing `#ifdef`.

File: drivers/coreaudio/audio_driver_coreaudio.mm [hunk 3]
```diff
@@ -405,13 +426,23 @@ Error AudioDriverCoreAudio::init_input_device() {
...
```
[the remark this answers is not in the document]
Comment (adamscott) [reply drivers/coreaudio/audio_driver_coreaudio.mm:434]: Fortunately, my tests still work after fixing `#ifdef`.

File: drivers/coreaudio/audio_driver_coreaudio.mm (hunk 2 above)
[the remark this answers is not in the document]
Comment (AThousandShips) [reply drivers/coreaudio/audio_driver_coreaudio.mm:124]: The rest of the code uses it why I suggested

File: drivers/coreaudio/audio_driver_coreaudio.mm (hunk 3 above)
[the remark this answers is not in the document]
Comment (bruvzg) [reply drivers/coreaudio/audio_driver_coreaudio.mm]: Default define value is `1`, so it should work the same.

**Review by Repiteo — APPROVED** [review]
Just needs a rebase & this'll be good to go!

**Comment by Repiteo** [comment]
Thanks!

**Pull request merged**